be/src/exprs/fpe.cpp

```diff
@@ -0,0 +1,197 @@
+    if (res != 0) {
+        free(out);
+        return Status::RuntimeError("ff1_encrypt failed");
+    }
+    ff1_ctx_destroy(ctx);
+
+    std::string result(out);
+    value = result;
+    free(out);
+
+    return Status::OK();
+}
+
+Status FPE::encrypt_num(const std::string& num_str,const std::string key, std::string& value) {
+    std::string num_flag;
+    if (num_str[0] == '-') {
+        num_flag = "-";
+    }
+
+    size_t dot_pos = num_str.find('.');
+    std::string int_part = num_str.substr(0, dot_pos);
+    Status status;
+    std::string encrypted_int_part;
+    status = encrypt(int_part, key, encrypted_int_part, DEFAULT_RADIX);
+    if (!status.ok()) {
+        return Status::RuntimeError("encrypt_num int_part failed");
+    }
+
+    std::string encrypted_dec_part;
+    if (dot_pos != std::string::npos) {
+        std::string dec_part = num_str.substr(dot_pos + 1);
+        dec_part = "0." + dec_part;
+        double dec_part_num = std::stod(dec_part);
+        dec_part_num = dec_part_num * EXPANDED;
+        auto dec_int_part = static_cast<long long>(dec_part_num);
+        status = encrypt(std::to_string(dec_int_part), key, encrypted_dec_part, DEFAULT_RADIX);
+        if (!status.ok()) {
+            return Status::RuntimeError("encrypt_num dec_int_part failed");
+        }
+        encrypted_dec_part = "." + encrypted_dec_part + FIXED_NUM;
+    }
+
+    value = num_flag + FIXED_NUM + encrypted_int_part + encrypted_dec_part;
+
+    return Status::OK();
+}
+
+Status FPE::decrypt(const std::string& num_str, const std::string key, std::string& value, int radix= 10) {
+    int num_str_length = num_str.length();
+    int fpe_key_length = key.length();
+
+    std::vector<uint8_t> fpe_key(fpe_key_length);
+    for (size_t i = 0; i < fpe_key_length; ++i) {
+        fpe_key[i] = static_cast<uint8_t>(key[i]);
+    }
+
+    struct ff1_ctx* ctx;
+    char* out = (char *) calloc(num_str_length + 1, sizeof(char));
+    if (out == NULL) {
+        return Status::MemoryAllocFailed("encrypt calloc memory Failed");
+    }
+
+    int res = ff1_ctx_create(&ctx, fpe_key.data(),
+                             fpe_key_length, TWEAK, sizeof(TWEAK),
+                             0, SIZE_MAX,
+                             radix);
+    if (res != 0) {
+        free(out);
+        return Status::RuntimeError("ff1_ctx_create failed");
+    }
+    res = ff1_decrypt(ctx, out, num_str.c_str(), NULL, 0);
+    if (res != 0) {
+        free(out);
+        return Status::RuntimeError("ff1_encrypt failed");
+    }
+    ff1_ctx_destroy(ctx);
+
+    std::string result(out);
+    value = result;
+    free(out);
+
+    return Status::OK();
+}
+
+Status FPE::decrypt_num(const std::string& num_str, const std::string key, std::string& value) {
+    std::string num_flag;
+    std::string decrypt_num_str = num_str;
+    if (decrypt_num_str[0] == '-') {
+        num_flag = "-";
+        decrypt_num_str = decrypt_num_str.substr(1);
+    }
```

Comment:
The string is copied too many times in `std:string::=`, `std::string::+`, `std::string::substring`.
Maybe we could give `value` enough room at first and perform all the char modification on the `values` to avoid copying string?